applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_cr_beam_element_3D2N.cpp

```diff
@@ -0,0 +1,287 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:     BSD License
+//           license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Martin Fusseder, https://github.com/MFusseder
+//
+
+
+#include "adjoint_finite_difference_cr_beam_element_3D2N.h"
+#include "structural_mechanics_application_variables.h"
+#include "custom_response_functions/response_utilities/response_data.h"
+#include "includes/checks.h"
+
+
+namespace Kratos
+{
+
+AdjointFiniteDifferenceCrBeamElement::AdjointFiniteDifferenceCrBeamElement(Element::Pointer pPrimalElement)
+    : AdjointFiniteDifferencingBaseElement(pPrimalElement)
+{
+}
+
+AdjointFiniteDifferenceCrBeamElement::~AdjointFiniteDifferenceCrBeamElement()
+{
+}
+
+Element::Pointer AdjointFiniteDifferenceCrBeamElement::Create(Element::Pointer pPrimalElement) const
+{
+    return Kratos::make_shared<AdjointFiniteDifferenceCrBeamElement>(pPrimalElement);
+}
+
+void AdjointFiniteDifferenceCrBeamElement::EquationIdVector(EquationIdVectorType& rResult,
+    ProcessInfo& rCurrentProcessInfo) {
+
+    const int number_of_nodes = this->GetGeometry().PointsNumber();
+    const int dimension = this->GetGeometry().WorkingSpaceDimension();
+    const unsigned int local_size = number_of_nodes * dimension * 2;
+
+    if (rResult.size() != local_size) rResult.resize(local_size);
+
+    GeometryType & geom = this->GetGeometry();
+
+    for (int i = 0; i < number_of_nodes; ++i)
+    {
+        int index = i * number_of_nodes * dimension;
+        rResult[index]     = geom[i].GetDof(ADJOINT_DISPLACEMENT_X).EquationId();
+        rResult[index + 1] = geom[i].GetDof(ADJOINT_DISPLACEMENT_Y).EquationId();
+        rResult[index + 2] = geom[i].GetDof(ADJOINT_DISPLACEMENT_Z).EquationId();
+
+        rResult[index + 3] = geom[i].GetDof(ADJOINT_ROTATION_X).EquationId();
+        rResult[index + 4] = geom[i].GetDof(ADJOINT_ROTATION_Y).EquationId();
+        rResult[index + 5] = geom[i].GetDof(ADJOINT_ROTATION_Z).EquationId();
+    }
+
+}
+
+void AdjointFiniteDifferenceCrBeamElement::GetDofList(DofsVectorType& rElementalDofList,
+    ProcessInfo& rCurrentProcessInfo) {
+
+    GeometryType & geom = this->GetGeometry();
+
+    const int number_of_nodes = geom.PointsNumber();
+    const int dimension = geom.WorkingSpaceDimension();
+    const unsigned int local_size = number_of_nodes * dimension * 2;
+
+    if (rElementalDofList.size() != local_size) {
+        rElementalDofList.resize(local_size);
+    }
+
+    for (int i = 0; i < number_of_nodes; ++i)
+    {
+        int index = i * number_of_nodes * dimension;
```

Comment:
This also seems wrong (copy paste from above). Can you just use the base class method?

Reply:
Done